File: BaseBackend/src/main/java/com/common/controller/AuthController.java
package com.common.controller;

import com.common.dto.request.AuthRequest;
import com.common.dto.request.RefreshTokenRequest;
import com.common.dto.request.RegisterRequest;
import com.common.dto.response.AuthResponse;
import com.common.dto.response.UserDTO;
import com.common.dto.response.WrapperApiResponse;
import com.common.service.MessageService;
import com.common.service.impl.AuthService;
import com.common.utilities.Utils;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * Controller xử lý các API xác thực Auth (Đăng ký, Đăng nhập, Refresh Token, Đăng xuất, User Info).
 *
 * @created_at 2026-07-31
 * @author txhoan
 */
@RestController
@RequestMapping("${app.api.auth-prefix:${server.servlet.context-path:/api/v1}/auth}")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    @PostMapping("/register")
    public ResponseEntity<WrapperApiResponse> register(@RequestBody @Valid RegisterRequest request) {
        AuthResponse response = authService.register(request);
        return Utils.returnResponse(org.springframework.http.HttpStatus.CREATED, MessageService.getMessage("controller.create_success"), response);
    }

    @PostMapping("/login")
    public ResponseEntity<WrapperApiResponse> login(@RequestBody @Valid AuthRequest request) {
        AuthResponse response = authService.login(request);
        return Utils.returnOkResponse(MessageService.getMessage("controller.get_success"), response);
    }

    @PostMapping("/refresh")
    public ResponseEntity<WrapperApiResponse> refreshToken(@RequestBody @Valid RefreshTokenRequest request) {
        AuthResponse response = authService.refreshToken(request);
        return Utils.returnOkResponse(MessageService.getMessage("controller.get_success"), response);
    }

    @PostMapping("/logout")
    public ResponseEntity<WrapperApiResponse> logout(@RequestHeader(value = "Authorization", required = false) String authHeader) {
        authService.logout(authHeader);
        return Utils.returnOkResponse(MessageService.getMessage("controller.delete_success"), null);
    }

    @GetMapping("/me")
    public ResponseEntity<WrapperApiResponse> getCurrentUser(Authentication authentication) {
        if (authentication == null || !authentication.isAuthenticated()) {
            return Utils.returnOkResponse(MessageService.getMessage("controller.get_success"), null);
        }
        UserDTO userDTO = authService.getCurrentUser(authentication.getName());
        return Utils.returnOkResponse(MessageService.getMessage("controller.get_success"), userDTO);
    }
}

File: be/.Net/Swe.API/Controllers/AuthController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Swe.BL.DTO.Request;
using Swe.BL.DTO.Response;
using Swe.BL.Service;
using Swe.Common.Model;

namespace Swe.API.Controllers;

/// <summary>
/// Controller xử lý các API xác thực Auth (Đăng ký, Đăng nhập, Refresh Token, Đăng xuất, User Info).
/// </summary>
/// <created_at>2026-08-03</created_at>
/// <author>txhoan</author>
[ApiController]
[Route("api/v1/auth")]
public class AuthController(AuthService authService) : ControllerBase
{
    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        try
        {
            var response = await authService.RegisterAsync(request);
            return StatusCode(201, new WrapperApiResponse(201, "Đăng ký tài khoản thành công", response));
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new WrapperApiResponse(400, ex.Message, null));
        }
        catch (Exception ex)
        {
            return StatusCode(500, new WrapperApiResponse(500, "Đã xảy ra lỗi hệ thống: " + ex.Message, null));
        }
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] AuthRequest request)
    {
        try
        {
            var response = await authService.LoginAsync(request);
            return Ok(new WrapperApiResponse(200, "Đăng nhập thành công", response));
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new WrapperApiResponse(400, ex.Message, null));
        }
        catch (Exception ex)
        {
            return StatusCode(500, new WrapperApiResponse(500, "Đã xảy ra lỗi hệ thống: " + ex.Message, null));
        }
    }

    [HttpPost("refresh")]
    public async Task<IActionResult> Refresh([FromBody] RefreshTokenRequest request)
    {
        try
        {
            var response = await authService.RefreshTokenAsync(request);
            return Ok(new WrapperApiResponse(200, "Làm mới token thành công", response));
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new WrapperApiResponse(400, ex.Message, null));
        }
        catch (Exception ex)
        {
            return StatusCode(500, new WrapperApiResponse(500, "Đã xảy ra lỗi hệ thống: " + ex.Message, null));
        }
    }

    [Authorize]
    [HttpPost("logout")]
    public IActionResult Logout()
    {
        // Phía client chỉ cần xóa token. Ở đây trả về thành công.
        return Ok(new WrapperApiResponse(200, "Đăng xuất thành công", null));
    }

    [Authorize]
    [HttpGet("me")]
    public async Task<IActionResult> GetCurrentUser()
    {
        var username = User.Identity?.Name ?? User.FindFirst(ClaimTypes.Name)?.Value;
        if (string.IsNullOrEmpty(username))
        {
            return Unauthorized(new WrapperApiResponse(401, "Chưa xác thực người dùng", null));
        }

        try
        {
            var userDto = await authService.GetCurrentUserAsync(username);
            return Ok(new WrapperApiResponse(200, "Lấy thông tin người dùng thành công", userDto));
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new WrapperApiResponse(400, ex.Message, null));
        }
        catch (Exception ex)
        {
            return StatusCode(500, new WrapperApiResponse(500, "Đã xảy ra lỗi hệ thống: " + ex.Message, null));
        }
    }
}
